hphp/util/mutex.h

```diff
@@ -326,6 +328,104 @@ struct RankedCHM : tbb::concurrent_hash_map<K, V, H> {
   bool erase(const K& k)                     { return RawCHM::erase(k); }
 };
 
+/*
+ * Need to subclass tbb::chm rather than RankedCHM so we
+ * have access to the protected my_hash_compare member
+ * for getting hashcodes to inject into const SmallLock locks[N].
+ * This is necessary for the erase case where we only have an accessor
+ * with a hashcode_t already inside, and need to be consistent
+ * with how we may have produced a hashcode_t earlier for the same key.
+ */
+#define DEFAULT_CHM_LOCK_BUCKET_COUNT 137
+template<typename K, typename V, typename H=K, Rank R=RankUnranked, uint32_t N=DEFAULT_CHM_LOCK_BUCKET_COUNT>
+struct RankedBucketCHM : tbb::concurrent_hash_map<K, V, H> {
+private:
+  typedef tbb::concurrent_hash_map<K, V, H> RawCHM;
+
+  /*
+   * InternalSmallLock is a union to avoid issues with default initialization of an
+   * array or container of SmallLocks - base_int is never actually used, only lock
+   */
+  typedef union { uint32_t base_int; mutable SmallLock lock; } InternalSmallLock;
```

Comment:
Is it worth a `static_assert(sizeof(unit32_t) == sizeof(SmallLock))`?

Reply:
Sounds like a good way to ensure at compile time that this union always works!

Reply:
A vector will work but sadly not a vector of raw `SmallLocks` because their default constructor has been deleted. The constructor argument is a good point, and I'll be switching to that pattern.

Reply:
I had a look at `smalllocks.h`. I don't think there's any reason to rely on that. You don't have a particular size concern with only 137 objects (in the default case). Just use an array of `std::mutex`s.